File: 110-BalancedBinaryTree/110-BalancedBinaryTree.py
# Last updated: 8/21/2025, 1:11:18 PM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#for each node, find height of left and right subtree
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        if root is None:
            return True
        
        def dfs(curr):
            leftHeight = 0
            rightHeight = 0
            leftBalanced = True
            rightBalanced = True    
            if curr.left is not None:
                leftHeight, leftBalanced = dfs(curr.left)
            if curr.right is not None:
                rightHeight, rightBalanced = dfs(curr.right)
            
            retBalanced = leftBalanced and rightBalanced and (abs(leftHeight-rightHeight)<=1)
            return 1 + max(leftHeight, rightHeight), retBalanced
        
        _, balanced = dfs(root)

        return balanced

File: 110-BalancedBinaryTree/110-BalancedBinaryTree.cpp
// Last updated: 8/21/2025, 1:50:21 PM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

//DFS algorithm search, want to keep

#include <cmath>

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        
        auto [dummy, isBalanced] = dfs(root);

        return isBalanced;

    }

    std::pair<int, bool> dfs(TreeNode* root){
        if (root==NULL){
            return std::make_pair(0,true);
        }
        

        auto [leftHeight, leftBalanced] = dfs(root->left);
        auto [rightHeight, rightBalanced] = dfs(root->right);

        bool balanced = leftBalanced && rightBalanced && (std::abs(leftHeight-rightHeight)<=1);
        
        int retHeight = std::max(leftHeight, rightHeight)+1;

        return std::make_pair(retHeight, balanced);
    }

};